code/datums/elements/decal_element.dm

```diff
@@ -0,0 +1,178 @@
+// NOTE:
+// This is an incredibly piecemeal port of /tg/'s decal element.
+// It does not include several pieces of functionality that exist in /tg/.
+//
+// Namely:
+// - It does not support smoothing decals
+// - It does not send a signal when a decal is detached (used for trapdoors on /tg/)
+// - It does not support custom plane configuration as this behavior seems primarily concerned with multi-z
+
+/datum/element/decal
+	element_flags = ELEMENT_BESPOKE|ELEMENT_DETACH_ON_HOST_DESTROY
+	argument_hash_start_idx = 2
+	/// Whether this decal can be cleaned.
+	var/cleanable
+	/// A description this decal appends to the target's examine message.
+	var/description
+	/// If true this was initialized with no set direction - will follow the parent dir.
+	var/directional
+	/// The base icon state that this decal was initialized with.
+	var/base_icon_state
+	/// The overlay applied by this decal to the target.
+	var/mutable_appearance/pic
+
+/// Remove old decals and apply new decals after rotation as necessary
+/datum/controller/subsystem/processing/dcs/proc/rotate_decals(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+
+	if(old_dir == new_dir)
+		return
+
+	var/list/datum/element/decal/old_decals = list() //instances
+	SEND_SIGNAL(source, COMSIG_ATOM_DECALS_ROTATING, old_decals)
+
+	if(!length(old_decals))
+		UnregisterSignal(source, COMSIG_ATOM_DIR_CHANGE)
+		return
+
+	var/list/resulting_decals_params = list() // param lists
+	for(var/datum/element/decal/rotating as anything in old_decals)
+		resulting_decals_params += list(rotating.get_rotated_parameters(old_dir,new_dir))
+
+	//Instead we could generate ids and only remove duplicates to save on churn on four-corners symmetry ?
+	for(var/datum/element/decal/decal in old_decals)
+		decal.Detach(source)
+
+	for(var/result in resulting_decals_params)
+		source.AddElement(/datum/element/decal, result["icon"], result["icon_state"], result["dir"], result["layer"], result["alpha"], result["color"], result["cleanable"], result["desc"])
+
+/datum/element/decal/proc/get_rotated_parameters(old_dir,new_dir)
+	var/rotation = 0
+	if(directional) //Even when the dirs are the same rotation is coming out as not 0 for some reason
+		rotation = SIMPLIFY_DEGREES(dir2angle(new_dir)-dir2angle(old_dir))
+		new_dir = turn(pic.dir,-rotation)
+	return list(
+		"icon" = pic.icon,
+		"icon_state" = base_icon_state,
+		"dir" = new_dir,
+		"plane" = pic.plane,
+		"layer" = pic.layer,
+		"alpha" = pic.alpha,
+		"color" = pic.color,
+		"cleanable" = cleanable,
+		"desc" = description
+	)
+
+/datum/element/decal/Attach(atom/target, _icon, _icon_state, _dir, _layer=TURF_LAYER, _alpha=255, _color, _cleanable=FALSE, _description, mutable_appearance/_pic)
```

Comment:
Also, might be nice to move attach directly under the typedef since it functions more or less as the initializer

Reply:
Good call, fixed.